vespa-cloud/vespa-documentation-search/src/main/java/ai/vespa/cloud/docsearch/OutLinksDocumentProcessor.java

```diff
@@ -0,0 +1,122 @@
+import com.yahoo.documentapi.messagebus.MessageBusDocumentAccess;
+
+import java.nio.file.Path;
+import java.util.Collections;
+import java.util.Set;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+/**
+ * Extract unique file-links from this document, and add as an in-link in the target document.
+ * Use async Document API to set in-links, fire-and-forget.
+ * This is an example Document Processor / use of Document API - and to be tested in a SystemTest.
+ */
+public class OutLinksDocumentProcessor extends DocumentProcessor {
+    private static final Logger logger = Logger.getLogger(OutLinksDocumentProcessor.class.getName());
+
+    static final String DOC_DOCUMENT_TYPE   = "doc";
+    static final String PATH_FIELD_NAME     = "path";
+    static final String OUTLINKS_FIELD_NAME = "outlinks";
+    static final String INLINKS_FIELD_NAME  = "inlinks";
+
+    private final MessageBusDocumentAccess access = new MessageBusDocumentAccess();
+    private final AsyncSession asyncSession = access.createAsyncSession(new AsyncParameters().setResponseHandler(new RespHandler()));
+
+    class RespHandler implements ResponseHandler {
+        @Override
+        public void handleResponse(Response response) {
+            //logger.info("In handleResponse");
+        }
+    }
+
+    @Override
+    public Progress process(Processing processing) {
+        //logger.info("In process");
+        for (DocumentOperation op : processing.getDocumentOperations()) {
+            if (op instanceof DocumentPut) {
+                DocumentPut put = (DocumentPut) op;
+                Document document = put.getDocument();
+                if (document.getDataType().isA(DOC_DOCUMENT_TYPE)) {
+                    String myPath = document.getFieldValue(PATH_FIELD_NAME).toString();
+
+                    Set<String> docsLinkedFromThis = canonicalizeLinks(Path.of(myPath),
+                            onlyFileLinks(removeLinkFragment(document)));
+
+                    Array<StringFieldValue> sfv = new Array<>(DataType.getArray(DataType.STRING));
+                    for (String link : docsLinkedFromThis) {
+                        sfv.add(new StringFieldValue(link));
+                    }
+                    document.setFieldValue(OUTLINKS_FIELD_NAME, sfv);
+
+                    if (docsLinkedFromThis.size() > 0) {
+                        addInLinks(docsLinkedFromThis, myPath, document.getDataType());
+                    }
+                }
+            }
+        }
+        return Progress.DONE;
+    }
+
+    @SuppressWarnings("unchecked")
+    private void addInLinks(Set<String> targetDocs, String myPath, DocumentType docType) {
+        WeightedSet<StringFieldValue> wset = new WeightedSet<>(DataType.getWeightedSet(DataType.STRING));
+        wset.put(new StringFieldValue(myPath), 1);
+        for (String targetDoc : targetDocs) {
+            String targetID = "id:open:" + DOC_DOCUMENT_TYPE + "::open" + targetDoc;
```

Comment:
Intentional `open` as start of user specified?

Reply:
yes - that is legacy - since then I use the namespace (open there, too), but did not yet remove from the user specified. can remove later, it is a larger change as the jekyll generator must use different IDs etc